src/ripple/app/paths/PathRequest.cpp

```diff
@@ -175,70 +166,69 @@ void PathRequest::updateComplete ()
 bool PathRequest::isValid (RippleLineCache::ref crCache)
 {
     ScopedLockType sl (mLock);
-    bValid = raSrcAccount && raDstAccount &&
-            saDstAmount > zero;
-    auto const& lrLedger = crCache->getLedger ();
+    if (! raSrcAccount || ! raDstAccount)
+        return false;
 
-    if (bValid)
+    if (! convert_all_ && (saSendMax || saDstAmount <= zero))
     {
-        if (! crCache->getLedger()->exists(
-                keylet::account(*raSrcAccount)))
-        {
-            // no source account
-            bValid = false;
-            jvStatus = rpcError (rpcSRC_ACT_NOT_FOUND);
-        }
+        // if send max specified, dst amt must be -1
+        jvStatus = rpcError(rpcDST_AMT_MALFORMED);
+        return false;
     }
 
-    if (bValid)
+    if (! crCache->getLedger()->exists(
+        keylet::account(*raSrcAccount)))
     {
-        auto const sleDest = crCache->getLedger()->read(
-            keylet::account(*raDstAccount));
+        // source account does not exist
+        jvStatus = rpcError (rpcSRC_ACT_NOT_FOUND);
+        return false;
+    }
 
-        Json::Value& jvDestCur =
-                (jvStatus[jss::destination_currencies] = Json::arrayValue);
+    auto const& lrLedger = crCache->getLedger();
+    auto const sleDest = crCache->getLedger()->read(
+        keylet::account(*raDstAccount));
 
-        if (!sleDest)
-        {
-            // no destination account
-            jvDestCur.append (Json::Value ("XRP"));
+    Json::Value& jvDestCur =
+        (jvStatus[jss::destination_currencies] = Json::arrayValue);
 
-            if (!saDstAmount.native ())
-            {
-                // only XRP can be send to a non-existent account
-                bValid = false;
-                jvStatus = rpcError (rpcACT_NOT_FOUND);
-            }
-            else if (saDstAmount < STAmount (lrLedger->fees().accountReserve (0)))
-            {
-                // payment must meet reserve
-                bValid = false;
-                jvStatus = rpcError (rpcDST_AMT_MALFORMED);
-            }
+    if (! sleDest)
+    {
+        // no destination account
+        jvDestCur.append (Json::Value ("XRP"));
+        if (! saDstAmount.native ())
+        {
+            // only XRP can be send to a non-existent account
```

Comment:
But this sort of comment as to why we're doing this is key to understanding this tricky business logic!